lib/std/crypto/argon2.zig

```diff
@@ -0,0 +1,1119 @@
+    var b2 = hasher.init(.{});
+
+    var buffer: [Blake2b512.digest_length]u8 = undefined;
+    mem.writeIntLittle(u32, buffer[0..4], @intCast(u32, out.len));
+    b2.update(buffer[0..4]);
+    b2.update(in);
+
+    if (out.len <= Blake2b512.digest_length) {
+        b2.final(out[0..hasher.digest_length]);
+        return;
+    }
+
+    std.debug.assert(out.len % 32 == 0);
+
+    b2.final(buffer[0..hasher.digest_length]);
+    b2 = hasher.init(.{});
+    mem.copy(u8, out, buffer[0..32]);
+    var out_slice = out[32..];
+    while (out_slice.len > Blake2b512.digest_length) : ({
+        out_slice = out_slice[32..];
+        b2 = hasher.init(.{});
+    }) {
+        b2.update(&buffer);
+        b2.final(buffer[0..hasher.digest_length]);
+        mem.copy(u8, out_slice, buffer[0..32]);
+    }
+
+    b2.update(&buffer);
+    b2.final(out_slice[0..hasher.digest_length]);
+}
+
+fn initBlocks(
+    blocks: *Blocks,
+    h0: *H0,
+    memory: u32,
+    threads: u32,
+) void {
+    var block0: [1024]u8 = undefined;
+    var lane: u32 = 0;
+    while (lane < threads) : (lane += 1) {
+        const j = lane * (memory / threads);
+        mem.writeIntLittle(u32, h0[Blake2b512.digest_length + 4 ..][0..4], lane);
+
+        mem.writeIntLittle(u32, h0[Blake2b512.digest_length..][0..4], 0);
+        blake2bHash(&block0, h0, Blake2b512);
+        for (blocks.items[j + 0]) |*v, i| {
+            v.* = mem.readIntLittle(u64, block0[i * 8 ..][0..8]);
+        }
+
+        mem.writeIntLittle(u32, h0[Blake2b512.digest_length..][0..4], 1);
+        blake2bHash(&block0, h0, Blake2b512);
+        for (blocks.items[j + 1]) |*v, i| {
+            v.* = mem.readIntLittle(u64, block0[i * 8 ..][0..8]);
+        }
+    }
+}
+
+fn processBlocks(
+    blocks: *Blocks,
+    time: u32,
+    memory: u32,
+    threads: u32,
+    mode: Mode,
+) Thread.SpawnError!void {
+    const lanes = memory / threads;
+    const segments = lanes / sync_points;
+
+    if (builtin.single_threaded or threads == 1) {
+        processBlocksSt(blocks, time, memory, threads, mode, lanes, segments);
+    } else {
+        try processBlocksMt(blocks, time, memory, threads, mode, lanes, segments);
+    }
+}
+
+fn processBlocksSt(
+    blocks: *Blocks,
+    time: u32,
+    memory: u32,
+    threads: u32,
+    mode: Mode,
+    lanes: u32,
+    segments: u32,
+) void {
+    var n: u32 = 0;
+    while (n < time) : (n += 1) {
+        var slice: u32 = 0;
+        while (slice < sync_points) : (slice += 1) {
+            var lane: u32 = 0;
+            while (lane < threads) : (lane += 1) {
+                processSegment(blocks, time, memory, threads, mode, lanes, segments, n, slice, lane);
+            }
+        }
+    }
+}
+
+fn processBlocksMt(
+    blocks: *Blocks,
+    time: u32,
+    memory: u32,
+    threads: u32,
+    mode: Mode,
+    lanes: u32,
+    segments: u32,
+) Thread.SpawnError!void {
+    var threads_arr: [256]Thread = undefined;
```

Comment:
If no more than 256 threads are allowed, maybe `threads` should be a `u8`.

Reply:
Changed to `u24` from RFC.